providers/adfs.go

```diff
@@ -0,0 +1,132 @@
+package providers
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"net/url"
+	"time"
+
+	"github.com/oauth2-proxy/oauth2-proxy/v7/pkg/apis/sessions"
+	"github.com/oauth2-proxy/oauth2-proxy/v7/pkg/requests"
+)
+
+// ADFSProvider represents an ADFS based Identity Provider
+type ADFSProvider struct {
+	*OIDCProvider
+}
+
+var _ Provider = (*OIDCProvider)(nil)
+
+const (
+	adfsProviderName = "Adfs"
+	adfsDefaultScope = "openid"
+)
+
+// NewADFSProvider initiates a new AdfsProvider
+func NewADFSProvider(p *OIDCProvider) *ADFSProvider {
+
+	p.ProviderData.setProviderDefaults(providerDefaults{
+		name:  adfsProviderName,
+		scope: adfsDefaultScope,
+	})
+	return &ADFSProvider{OIDCProvider: p}
+}
+
+// GetLoginURL overrides GetLoginURL to add Adfs parameters
+func (p *ADFSProvider) GetLoginURL(redirectURI, state string) string {
+	extraParams := url.Values{}
+	if p.ProtectedResource != nil && p.ProtectedResource.String() != "" {
+		extraParams.Add("resource", p.ProtectedResource.String())
+	}
+	a := makeLoginURL(p.ProviderData, redirectURI, state, extraParams)
+	return a.String()
+}
+
+// Redeem exchanges the OAuth2 authentication token for an Access\ID tokens
+func (p *ADFSProvider) Redeem(ctx context.Context, redirectURL, code string) (s *sessions.SessionState, err error) {
+	if code == "" {
+		return nil, ErrMissingCode
+	}
+	clientSecret, err := p.GetClientSecret()
+	if err != nil {
+		return nil, err
+	}
+
+	var jsonResponse struct {
+		AccessToken  string `json:"access_token"`
+		RefreshToken string `json:"refresh_token"`
+		ExpiresIn    int64  `json:"expires_in"`
+		IDToken      string `json:"id_token"`
+	}
+
+	params := url.Values{}
+	params.Add("client_id", p.ClientID)
+	params.Add("client_secret", clientSecret)
+	if p.ProtectedResource != nil && p.ProtectedResource.String() != "" {
+		params.Add("resource", p.ProtectedResource.String())
+	}
+	params.Add("grant_type", "authorization_code")
+	params.Add("code", code)
+	params.Add("redirect_uri", redirectURL)
+
+	err = requests.New(p.RedeemURL.String()).
+		WithContext(ctx).
+		WithMethod("POST").
+		WithBody(bytes.NewBufferString(params.Encode())).
+		SetHeader("Content-Type", "application/x-www-form-urlencoded").
+		Do().
+		UnmarshalInto(&jsonResponse)
+	if err != nil {
+		return
+	}
+
+	created := time.Now()
+	expires := time.Unix(jsonResponse.ExpiresIn, 0)
```

Comment:
So the OAuth response for this field from the `token_endpoint` should match up with the Access Token lifespan. But you also have an ID Token in the response to this about.

You might want to make sure the ID Token also has this Expiry.

In OIDC, since the ID Token is the main token we care about, other OIDC based providers set the `session.ExpiresOn` from the `idToken.Expiry` explicitly instead of going off the OAuth2 `expires_in` field.